<html lang="en">
<head>
  <h1>Dog and Owner Characteristics Predict Training Success Dataset</h1>
  <meta charset="utf-8">
</head>
<body>

<div class="alert alert-info">
<strong>Part 1: </strong> Selecting, Extracting, Transforming, Analyzing, and Visualizing Data
</div>

</body>
</html>

## Part 1 - Questions

### *Question 1:* 

Does a dog's owner's extroversion and stability affect their nervousness and disobedience?

* Who (population): Dogs and their Owners
* What (subject, discipline): Training Success
* Where (location): USA
* When (snapshot, longitudinal): July 2020 - February 2021
* How much data do you need to do the analysis/work: around 50 observations

### *Question 2:*

Does a dog's owner's extroversion and stability affect the success of the training?

* Who (population): Dogs and their Owners
* What (subject, discipline): Training Success
* Where (location): USA
* When (snapshot, longitudinal): July 2020 - February 2021
* How much data do you need to do the analysis/work: around 50 observations

### *Who Might Collect Relevant Data / What Articles or Publications Cite a Relevant Data Set?*

Open Science Framework (OSF)

## Part 2 - Selecting a Data Set, Adding Documentation

### *Dataset Documentation*
1. Name / Title: Dog and owner characteristics predict training success
2. Link to Data: https://osf.io/3p5vx/
3. Source / Origin: 
	* Author or Creator: Jeffrey R Stevens London Wolff
	* Publication Date: 2020-07-10
	* Publisher: Open Science Framework (OSF)
	* Version or Data Accessed: stevens_etal_2020_data1
4. License: CC-By Attribution 4.0 International
5. Can You Use this Data Set for Your Intended Use Case? Yes

### *Format*
#### 1. Overview

* Format: CSV 
* Size: 16KB 
* Number of Records: 99 



#### 2. Sample of Data

In [ ]:
# a few lines of data from the actual file

with open('../data/raw/stevens_etal_2020_data1.csv', 'r') as f:
   headers = [h.strip() for h in next(f).split(',')]
   id_i = headers.index('id')
   owner_gender_i = headers.index('owner_gender')
   owner_extraversion_i = headers.index('personality_extraversion_score')
   owner_stability_i = headers.index('personality_stability_score')
   dog_gender_i = headers.index('dog_sex')
   dog_disobedient_i = headers.index('dog_behavior_bennett_disobedient_score')
   dog_nervous_i = headers.index('dog_behavior_bennett_nervous_score')
   cgc_result_i = headers.index('cgc_test')
   for line in f:
        line_pieces = line.split(',')
        # showing only a few lines (20 lines)
        if int(line_pieces[id_i]) <= 20: 
           print('ID: ', line_pieces[id_i])
           print('Owner Gender: ', line_pieces[owner_gender_i], '|' ,'Owner Extraversion Score: ', line_pieces[owner_extraversion_i], '|', 'Owner Stability Score: ', line_pieces[owner_stability_i])
           print('Dog Gender  : ', line_pieces[dog_gender_i], '|' ,'Dog Disobedient Score   : ', line_pieces[dog_disobedient_i], '|', 'Dog Nervous Score: ', line_pieces[dog_nervous_i])
           print('CGC Test Result: ', line_pieces[cgc_result_i])

ID:  1
Owner Gender:  Female | Owner Extraversion Score:  2 | Owner Stability Score:  5
Dog Gender  :  Male | Dog Disobedient Score   :  -0.05 | Dog Nervous Score:  0.14
CGC Test Result:  Pass

ID:  2
Owner Gender:  Female | Owner Extraversion Score:  3.5 | Owner Stability Score:  4
Dog Gender  :  Male | Dog Disobedient Score   :  -0.18 | Dog Nervous Score:  0.65
CGC Test Result:  Pass

ID:  3
Owner Gender:  NA | Owner Extraversion Score:  NA | Owner Stability Score:  NA
Dog Gender  :  NA | Dog Disobedient Score   :  NA | Dog Nervous Score:  NA
CGC Test Result:  NA

ID:  4
Owner Gender:  Female | Owner Extraversion Score:  3 | Owner Stability Score:  3.5
Dog Gender  :  Female | Dog Disobedient Score   :  0.15 | Dog Nervous Score:  -0.38
CGC Test Result:  Pass

ID:  5
Owner Gender:  NA | Owner Extraversion Score:  NA | Owner Stability Score:  NA
Dog Gender  :  NA | Dog Disobedient Score   :  NA | Dog Nervous Score:  NA
CGC Test Result:  NA

ID:  6
Owner Gender:  Female | Owner Extravers


#### 3. Fields or Column Headers

 There are 37 columns, only identifying a subset of 7 columns
* Field/Column 1: Owner Gender, str
* Field/Column 2: Owner Extraversion Score, int
* Field/Column 3: Owner Stability Score, int
* Field/Column 4: Dog Gender, str
* Field/Column 5: Dog Disobedient Score, int
* Field/Column 6: Dog Nervousness Score, int
* Field/Column 7: CGC Test Result, str

## Part 3 - Extract / Transform

* I will be using 7 columns: Owner Gender, Owner Extraversion Score, Owner Stability Score, Dog Gender, Dog Disobedient Score, Dog Nervousness Score, and CGC Test Results
* For the numeric data, I will use str.format() to format all the scores to have 3 places after the decimal point to consider calculations.
* For the categorical data, I will extract the data as is as all entries are normalized to 'Female/Male' or 'Pass/Fail'.
* I will use a generator to create an iterator that will iterate over the csv file, extracting the data needed. 

In [ ]:
# get the indices of the columns to extract
with open('../data/raw/stevens_etal_2020_data1.csv', 'r') as f:
   headers = [h.strip() for h in next(f).split(',')]
   id_i = headers.index('id')
   owner_gender_i = headers.index('owner_gender')
   owner_extraversion_i = headers.index('personality_extraversion_score')
   owner_stability_i = headers.index('personality_stability_score')
   dog_gender_i = headers.index('dog_sex')
   dog_disobedient_i = headers.index('dog_behavior_bennett_disobedient_score')
   dog_nervous_i = headers.index('dog_behavior_bennett_nervous_score')
   cgc_result_i = headers.index('cgc_test')
   
import csv
# read file using generator
def csv_reader(fp):
   file = open(fp, 'r')
   reader = csv.reader(file)
   next(reader)
   for row in file:
       line_list = []
       line_pieces = row.split(',')
       # extract only rows needed 
       line_list.append(int(line_pieces[id_i]))
       if 'NA' in line_pieces[owner_gender_i]:
           line_list.append(None)
       else: 
           line_list.append(line_pieces[owner_gender_i]) 
       if 'NA' in line_pieces[owner_extraversion_i]:
           line_list.append(None)
       else: 
           line_list.append(float(line_pieces[owner_extraversion_i])) 
       if 'NA' in line_pieces[owner_stability_i]:
           line_list.append(None)
       else: 
           line_list.append(float(line_pieces[owner_stability_i])) 
       if 'NA' in line_pieces[dog_gender_i]:
           line_list.append(None)
       else: 
           line_list.append(line_pieces[dog_gender_i]) 
       if 'NA' in line_pieces[dog_disobedient_i]:
           line_list.append(None)
       else: 
           line_list.append(float(line_pieces[dog_disobedient_i])) 
       if 'NA' in line_pieces[dog_nervous_i]:
           line_list.append(None)
       else: 
           line_list.append(float(line_pieces[dog_nervous_i])) 
       if 'NA' in line_pieces[cgc_result_i]:
           line_list.append(None)
       else: 
           line_list.append(line_pieces[cgc_result_i]) 
       yield line_list 
file_path = '../data/raw/stevens_etal_2020_data1.csv'
arr_data = []
for line in csv_reader(file_path): 
   arr_data.append(line)
# storing one num column (owner extraversion score) to a python list
num_data = []
rowCount = 0
for row in arr_data: 
   num_data.append(arr_data[rowCount][2])
   rowCount += 1
#print(num_data)
   #extract the 7 columns to work with
   #transform 4 columns to a Python data type
   #store data in a data type of choosing
   #clean data
   #use a generator

## Part 4 - Descriptive Statistics

### *Analysis on Numeric Data*

#### 1. Central Tendency

In [ ]:

# calculate mean
mean = 0
numOfNums = 0
meanCount = 0
# loop through all values of num_data and add up
for num in num_data: 
   if num_data[meanCount] != None: 
       mean += num_data[meanCount]
       numOfNums += 1
   meanCount += 1
mean = mean / numOfNums
print('Mean: ', mean)


Mean:  3.9761904761904763


#### 2. Dispersion

In [ ]:
# calculate range
lowest = num_data[0]
highest = 0
for num in num_data: 
  if num != None: 
      if num > highest: 
          highest = num
      if num < lowest: 
          lowest = num
range = highest - lowest
print('Range: ', range)


Range:  6.0


#### 3. Outliers


In [ ]:
print('Outliers: ', lowest, highest)


#### 4. Other


In [ ]:
print('Min: ', lowest, '| Max: ', highest)


###  *Analysis on Categorical Data*

#### 1. Frequency

In [ ]:
freqCount = {}
for element in num_data: 
   if element in freqCount: 
       freqCount[element] += 1
   else: 
       freqCount[element] = 1
print(freqCount)


#### 2. Unique Values

In [ ]:
value = 1
key_list = list(freqCount.keys())
val_list = list(freqCount.values())
if value in freqCount.values(): 
   position = val_list.index(value)
   print(key_list[position])
else: 
   print('No unique values exist.')


## Part 5 - Visualizations


In [ ]:
key_list = list(freqCount.keys())
val_list = list(freqCount.values())
import matplotlib.pyplot as plt
plt.scatter(key_list, val_list, color = 'blue')
plt.xlabel('Owner Extraversion Score')
plt.ylabel('Frequency of Score')
plt.title('Scatter Plot: Frequency of Owner Extraversion Scores')
plt.show() 


Scatter Plot: Frequency of Owner Extraversion Scores

The scatter plot displays the frequencies of the Extraversion Scores owners received. The graph shows that 3.5 is the most popular score received.


## Part 6 - Conclusion

The calculations and visualizations helped to answer my second question which was 'Does a dog's owner's extroversion and stability affect the success of the training?'
Based on the calculations, the extroversion scores did not vary by much and the range showed that there was a not a great difference between the min and max.
The graph showed that frequency of scores also did not vary; therefore, an owner's extroversion scores did not play a big part to determine the success of their dog's training.
